File: internal/registry/registry.go
package registry

import (
	"errors"
	"fmt"
	"sync"

	"github.com/Georgecane/openhost/internal/fabric"
	"github.com/Georgecane/openhost/internal/identity"
	"github.com/Georgecane/openhost/internal/participant"
)

var (
	ErrParticipantExists    = errors.New("participant already registered")
	ErrParticipantNotFound  = errors.New("participant not found")
	ErrParticipantNotActive = errors.New("participant is not active")
)

type Registry struct {
	mu           sync.RWMutex
	participants map[string]*participant.Participant
}

func New() *Registry {
	return &Registry{
		participants: make(map[string]*participant.Participant),
	}
}

func (r *Registry) Register(p *participant.Participant) error {
	if r == nil || p == nil {
		return ErrParticipantNotFound
	}

	snapshot := p.Snapshot()
	if err := snapshot.ID.Validate(); err != nil || snapshot.ID.Kind != identity.ParticipantKind {
		return fmt.Errorf("%w: participant identity must be valid", participant.ErrInvalidParticipant)
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.participants[snapshot.ID.ID]; exists {
		return ErrParticipantExists
	}
	r.participants[snapshot.ID.ID] = p
	return nil
}

func (r *Registry) Remove(id identity.Identity) error {
	if r == nil {
		return ErrParticipantNotFound
	}
	if err := id.Validate(); err != nil || id.Kind != identity.ParticipantKind {
		return fmt.Errorf("%w: participant identity must be valid", participant.ErrInvalidParticipant)
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.participants[id.ID]; !exists {
		return ErrParticipantNotFound
	}
	delete(r.participants, id.ID)
	return nil
}

func (r *Registry) Get(id identity.Identity) (*participant.Participant, error) {
	if r == nil {
		return nil, ErrParticipantNotFound
	}
	if err := id.Validate(); err != nil || id.Kind != identity.ParticipantKind {
		return nil, fmt.Errorf("%w: participant identity must be valid", participant.ErrInvalidParticipant)
	}

	r.mu.RLock()
	defer r.mu.RUnlock()

	p, exists := r.participants[id.ID]
	if !exists {
		return nil, ErrParticipantNotFound
	}
	return p, nil
}

func (r *Registry) Offers() []fabric.ResourceOffer {
	if r == nil {
		return nil
	}

	r.mu.RLock()
	defer r.mu.RUnlock()

	offers := make([]fabric.ResourceOffer, 0, len(r.participants))
	for id, p := range r.participants {
		resources, err := p.Offer()
		if err != nil {
			continue
		}
		offers = append(offers, fabric.ResourceOffer{
			ParticipantID: id,
			Resources:     resources,
		})
	}
	return offers
}

File: src/registry.rs
use crate::fabric::{Fabric, ResourceOffer};
use crate::identity::{Identity, Kind};
use crate::participant::{Participant, ParticipantError, State};
use std::collections::BTreeMap;
use std::sync::{Arc, RwLock};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RegistryError { Exists, NotFound, Invalid }

impl std::fmt::Display for RegistryError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { write!(f, "{self:?}") }
}
impl std::error::Error for RegistryError {}

#[derive(Default)]
pub struct Registry {
    participants: RwLock<BTreeMap<String, Arc<Participant>>>,
}

impl Registry {
    pub fn new() -> Self { Self::default() }

    pub fn register(&self, participant: Arc<Participant>) -> Result<(), RegistryError> {
        let id = participant.snapshot().id;
        if id.kind != Kind::Participant || id.validate().is_err() { return Err(RegistryError::Invalid); }
        let mut map = self.participants.write().expect("registry lock poisoned");
        if map.contains_key(&id.id) { return Err(RegistryError::Exists); }
        map.insert(id.id, participant);
        Ok(())
    }

    pub fn remove(&self, id: &Identity) -> Result<(), RegistryError> {
        if id.kind != Kind::Participant || id.validate().is_err() { return Err(RegistryError::Invalid); }
        if self.participants.write().expect("registry lock poisoned").remove(&id.id).is_none() {
            return Err(RegistryError::NotFound);
        }
        Ok(())
    }

    pub fn get(&self, id: &Identity) -> Result<Arc<Participant>, RegistryError> {
        if id.kind != Kind::Participant || id.validate().is_err() { return Err(RegistryError::Invalid); }
        self.participants.read().expect("registry lock poisoned").get(&id.id).cloned().ok_or(RegistryError::NotFound)
    }
}

impl Fabric for Registry {
    fn offers(&self) -> Vec<ResourceOffer> {
        self.participants.read().expect("registry lock poisoned").values().filter_map(|p| {
            match (p.snapshot().state, p.offer()) {
                (State::Active, Ok(resources)) => Some(ResourceOffer { participant_id: p.snapshot().id.id, resources }),
                _ => None,
            }
        }).collect()
    }
}

impl From<ParticipantError> for RegistryError {
    fn from(_: ParticipantError) -> Self { RegistryError::Invalid }
}
